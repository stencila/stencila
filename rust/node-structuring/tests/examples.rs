use std::{
    fs::{read_dir, read_to_string},
    path::PathBuf,
    str::FromStr,
};

use eyre::{OptionExt, Result, bail};
use insta::assert_json_snapshot;
use itertools::Itertools;

use stencila_codec::{Codec, StructuringOperation, StructuringOptions, stencila_format::Format};
use stencila_codec_json::JsonCodec;
use stencila_codec_markdown::MarkdownCodec;
use stencila_node_structuring::structuring;
use stencila_schema::{Node, Primitive};

/// Structure example Markdown or JSON files using the structuring operation
/// specified in the first three parts of their name, or multiple operations for
/// files with numeric prefixes.
///
/// Run using:
///
/// cargo insta test -p stencila-node-structuring
///
/// and then review snapshots with:
///
/// cargo insta review
#[tokio::test]
async fn examples() -> Result<()> {
    let dir = PathBuf::from(env!("CARGO_MANIFEST_DIR")).join("tests/examples");

    for entry in read_dir(&dir)?.flatten() {
        let path = entry.path();
        let format = Format::from_path(&path);

        let content = read_to_string(&path)?;

        let (mut node, ..) = match format {
            Format::Markdown => MarkdownCodec.from_str(&content, None).await?,
            Format::Json => JsonCodec.from_str(&content, None).await?,
            _ => bail!("Unsupported formatted"),
        };

        let id = path
            .file_stem()
            .map(|name| name.to_string_lossy().to_string())
            .ok_or_eyre("No file stem")?;

        let Node::Article(article) = &node else {
            bail!("Expected an article")
        };

        let operations = if let Some(Primitive::Array(ops)) = article
            .options
            .extra
            .as_ref()
            .and_then(|extra| extra.get("structuring").cloned())
        {
            ops.iter()
                .filter_map(|op| match op {
                    Primitive::String(op) => StructuringOperation::from_str(op).ok(),
                    _ => None,
                })
                .collect_vec()
        } else {
            // Fall back to operation from filename for single operation tests
            let op = id.split("-").take(3).collect_vec().join("-");
            if let Ok(operation) = StructuringOperation::from_str(&op) {
                vec![operation]
            } else {
                bail!("Unable to determine structuring operations for {}", path.display());
            }
        };

        structuring(&mut node, StructuringOptions::new(operations, []));

        assert_json_snapshot!(id, node);
    }

    Ok(())
}
