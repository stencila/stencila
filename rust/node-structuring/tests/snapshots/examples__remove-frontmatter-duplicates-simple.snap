---
source: rust/node-structuring/tests/examples.rs
expression: node
---
{
  "type": "Article",
  "doi": "10.1038/s41467-025-12345-6",
  "authors": [
    {
      "type": "Person",
      "orcid": "0000-0001-2345-6789",
      "affiliations": [
        {
          "type": "Organization",
          "name": "Department of Molecular Biology, University of Barcelona",
          "ror": "03yjgw361"
        }
      ],
      "familyNames": [
        "Rodriguez"
      ],
      "givenNames": [
        "Elena"
      ],
      "emails": [
        "elena.rodriguez@example.org"
      ]
    },
    {
      "type": "Person",
      "orcid": "0000-0002-3456-7890",
      "affiliations": [
        {
          "type": "Organization",
          "name": "Institute for Cell Biology Research",
          "ror": "04xfq0f34"
        }
      ],
      "familyNames": [
        "Chen"
      ],
      "givenNames": [
        "Marcus"
      ],
      "emails": [
        "marcus.chen@example.org"
      ]
    }
  ],
  "title": [
    {
      "type": "Text",
      "value": {
        "string": "Understanding Molecular Mechanisms of Cell Division in Yeast"
      }
    }
  ],
  "frontmatter": "type: Article\ntitle: Understanding Molecular Mechanisms of Cell Division in Yeast\ndoi: 10.1038/s41467-025-12345-6\nauthors:\n- type: Person\n  givenNames:\n  - Elena\n  familyNames:\n  - Rodriguez\n  orcid: 0000-0001-2345-6789\n  email: elena.rodriguez@example.org\n  affiliations:\n  - type: Organization\n    name: Department of Molecular Biology, University of Barcelona\n    ror: 03yjgw361\n- type: Person\n  givenNames:\n  - Marcus\n  familyNames:\n  - Chen\n  orcid: 0000-0002-3456-7890\n  email: marcus.chen@example.org\n  affiliations:\n  - type: Organization\n    name: Institute for Cell Biology Research\n    ror: 04xfq0f34",
  "content": [
    {
      "type": "Heading",
      "level": 1,
      "content": [
        {
          "type": "Text",
          "value": {
            "string": "Introduction"
          }
        }
      ]
    },
    {
      "type": "Paragraph",
      "content": [
        {
          "type": "Text",
          "value": {
            "string": "Cell division is a fundamental process in all living organisms. This paper explores the molecular mechanisms underlying mitosis in Saccharomyces cerevisiae."
          }
        }
      ]
    },
    {
      "type": "Heading",
      "level": 1,
      "content": [
        {
          "type": "Text",
          "value": {
            "string": "Methods"
          }
        }
      ]
    },
    {
      "type": "Paragraph",
      "content": [
        {
          "type": "Text",
          "value": {
            "string": "We used fluorescence microscopy to observe cell division in real time."
          }
        }
      ]
    }
  ]
}
